p2panda-sync/src/test_protocols.rs

```diff
@@ -0,0 +1,376 @@
+                        .unwrap();
+
+                    sink.send(PingPongProtocolMessage::Pong).await?;
+                    debug!("pong message sent");
+                    break;
+                }
+                PingPongProtocolMessage::Pong => {
+                    return Err(SyncError::UnexpectedBehaviour(
+                        "unexpected Pong message received".to_string(),
+                    ));
+                }
+            }
+        }
+
+        sink.flush().await?;
+        app_tx.flush().await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+enum FailingProtocolMessage {
+    TopicQuery(SyncTestTopic),
+    Done,
+}
+
+/// A sync implementation which returns a mocked error.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub enum FailingProtocol {
+    /// A critical error is triggered inside `accept()` after sync messages have been
+    /// exchanged.
+    AcceptorFailsCritical,
+
+    /// A critical error is triggered inside `initiate()` after the handshake is complete.
+    InitiatorFailsCritical,
+
+    /// An unexpected behaviour error is triggered inside `initiate()` after the topic query has
+    /// been sent.
+    InitiatorFailsUnexpected,
+
+    /// An unexpected behaviour error is triggered inside `accept()` by sending the topic twice
+    /// from `initiate()`.
+    InitiatorSendsTopicTwice,
+
+    /// An unexpected behaviour error is triggered inside `initiate()` by sending a topic from
+    /// `accept()`.
+    AcceptorSendsTopic,
+
+    /// No errors are explicitly triggered; used for "happy path" test.
+    NoError,
+}
+
+#[async_trait]
+impl<'a> SyncProtocol<'a, SyncTestTopic> for FailingProtocol {
+    fn name(&self) -> &'static str {
+        "failing-protocol"
+    }
+
+    async fn initiate(
+        self: Arc<Self>,
+        topic: SyncTestTopic,
+        tx: Box<&'a mut (dyn AsyncWrite + Send + Unpin)>,
+        rx: Box<&'a mut (dyn AsyncRead + Send + Unpin)>,
+        mut app_tx: Box<
+            &'a mut (dyn Sink<FromSync<SyncTestTopic>, Error = SyncError> + Send + Unpin),
+        >,
+    ) -> Result<(), SyncError> {
+        let mut sink = into_cbor_sink(tx);
+        let mut stream = into_cbor_stream(rx);
+
+        sink.send(FailingProtocolMessage::TopicQuery(topic.clone()))
+            .await?;
+
+        // Simulate critical sync implementation bug by sending the topic a second time.
+        if let FailingProtocol::InitiatorSendsTopicTwice = *self {
+            sink.send(FailingProtocolMessage::TopicQuery(topic.clone()))
+                .await?;
+        }
+
+        // Simulate some critical error which occurred inside the sync session.
+        if let FailingProtocol::InitiatorFailsCritical = *self {
+            return Err(SyncError::Critical(
+                "something really bad happened in the initiator".to_string(),
+            ));
+        }
+
+        // Simulate unexpected behaviour (such as a broken pipe due to disconnection).
```

Comment:
This is the newly-introduced test protocol variant that we use to test the resync-after-any-error behaviour.